### ISSUE TITLE
TF 2.0: tf.function+dataset do not use GPU by default

### ISSUE THREAD

**Issue by dimitree54** [tensorflow/tensorflow#32138]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): tensorflow-gpu==2.0.0rc0 
- Python version: 3.7.4
- CUDA/cuDNN version: CUDA 10.0/ cuDNN 7.6
- GPU model and memory: NVIDIA GeForce GT 630, 2048Mb

**Describe the current behavior**
Wrapping iteration through the dataset with tf.function move computations from GPU to CPU.

**Describe the expected behavior**
I read in [tutorial](https://www.tensorflow.org/beta/guide/effective_tf2) that wrapping iteration through the dataset with tf.function should increase performance, but instead of it computations move from GPU to CPU and as result slow down. Is it correct behavior? 

I can fix this behavior by manual device placement, by I think that computations should be done on GPU by default.

**Code to reproduce the issue**
```python
import tensorflow as tf
import time

data = tf.Variable(initial_value=tf.zeros((1, 1000, 1000)))
dataset = tf.data.Dataset.from_tensor_slices(data).repeat(1000)

weights = tf.Variable(initial_value=tf.zeros((1000, 1000)))

result = tf.Variable(initial_value=tf.zeros((1000, 1000)))

def py_process():
    for sample in dataset:
        result.assign_add(tf.matmul(sample, weights))
    return result

@tf.function
def tf_process():
    for sample in dataset:
        result.assign_add(tf.matmul(sample, weights))
    return result

@tf.function
def tf_gpu_process():
    for sample in dataset:
        with tf.device('/device:GPU:0'):
            result.assign_add(tf.matmul(sample, weights))
    return result

t = time.time()
py_process()
print("py time", time.time() - t)

t = time.time()
tf_process()
print("tf time", time.time() - t)

t = time.time()
tf_gpu_process()
print("tf gpu time", time.time() - t)
```

Output:
```
py time 14.078931093215942
tf time 47.290977478027344
tf gpu time 13.702632665634155
```

**Other info / logs**
I am tracking GPU usage with GPU-Z program, here is screenshot:
![gpu-z](https://user-images.githubusercontent.com/39711437/64075084-56947a80-ccbc-11e9-9853-66555d2c5d06.gif)

**Comment by ravikyram** [comment]
I have tried on colab with TF version 2.0.0-rc0, 2.0.0-dev20190902 and was able to reproduce the issue.Please, find the [gist](https://colab.sandbox.google.com/gist/ravikyram/4eb507f8f61984e08f4b5a02bb955307/untitled146.ipynb) here. Thanks!

**Comment by kkimdev** [comment]
I'm looking at this. @jsimsa please let me know if you are aware of a similar issue.

**Comment by jsimsa** [comment]
Could you please re-run your program logging device placement `https://www.tensorflow.org/beta/guide/using_gpu#logging_device_placement?

This will provide information about the placement of each op. My expectation is that the `reduce` dataset op introduced by Autograph will be placed on CPU while the `assign_add` and `matmul` ops will be placed on GPU.

**Comment by dimitree54** [comment]
> Could you please re-run your program logging device placement `https://www.tensorflow.org/beta/guide/using_gpu#logging_device_placement?
> 
> This will provide information about the placement of each op. My expectation is that the `reduce` dataset op introduced by Autograph will be placed on CPU while the `assign_add` and `matmul` ops will be placed on GPU.

Here is full program log including device placement.
[log.txt](https://github.com/tensorflow/tensorflow/files/3575806/log.txt)

**Comment by jsimsa** [comment]
The log information confirms my expectation. The reduce dataset op is placed on CPU but the `assign_add` and `matmul` ops are placed on GPU.

**Comment by alextp** [comment]
It seems like the log is only showing the matmul op executed eagerly, not the matmul op executed inside the reducedataset op. That matmul seems to be running on the CPU.

**Comment by jsimsa** [comment]
@dimitree54 we came up with a two-part fix (tf.data + tf.autograph). The tf.data  part has been completed (https://github.com/tensorflow/tensorflow/commit/cc2f631baaa9d0d5e189d3c0b5fbfc4df110bf91#diff-b9d6ae3057dedddc9760f15603a3b041) and the tf.autograph part is worked on by @mdanatg .

**Comment by dimitree54** [comment]
@jsimsa Ok, thank you. Is the second part (with autograph) fixed already? In which version of tensorflow 2 this fix will be available?

**Comment by mdanatg** [comment]
Yes, the autograph workaround just landed, you should be able to try it out in the next nightly, and will be available in 2.1.

**Comment by mdanatg** [comment]
In the mean time, the following workaround might work as well: `for sample in iter(dataset)`. Writing it in that way uses different placement rules, for subtle reasons.
